### PR TITLE
Resolve symlinks when identifying configured SHELL

### PR DESCRIPTION
**Pull request by mdrohmann**
At least in Ubuntu, it is configurable what `/bin/sh` can point to.  By default it is `/bin/dash` (a POSIX compatible shell) which is a subset of `bash`.  (Maybe we should support it, or ensure that our bash scripts run with just POSIX commands.

I also changed the logging level from Error to Debug, as the message does not give us much actionable information.  It is just interesting to know what shells people are using.

 https://www.pivotaltracker.com/story/show/177207884

### CODE REVIEW

File: internal/subshell/subshell.go [hunk 1]
```diff
@@ -84,10 +86,20 @@ func New() SubShell {
...
 	}
 
-	logging.Debug("Detected SHELL: %s", binary)
+	// try to find the binary on the PATH
+	binaryPath, err := exec.LookPath(binary)
+	if err == nil {
+		// if we found it, resolve all symlinks, for many Linux distributions the SHELL is "sh" but symlinked to a different default shell like bash or zsh
+		resolved, err := fileutils.ResolvePath(binaryPath)
+		if err != nil {
+			logging.Debug("Failed to resolve path to shell binary %s", binaryPath)
```
Comment (MDrakos) [inline internal/subshell/subshell.go:95]: Nit: We're losing the error here and it could be helpful to have this in the logs

**Review by MDrakos — APPROVED** [review]

**Review by MDrakos — DISMISSED** [review]

File: internal/subshell/subshell.go [hunk 2]
```diff
@@ -108,7 +120,7 @@ func New() SubShell {
 	case "cmd":
 		subs = &cmd.SubShell{}
 	default:
-		logging.Errorf("Unsupported shell: %s, defaulting to OS default.", name)
+		logging.Debug("Unsupported shell: %s, defaulting to OS default.", name)
```
Comment (mdrohmann) [inline internal/subshell/subshell.go:123]: I wonder if it is worthwhile creating an analytics event for this one.  So, we can find out what the most prevalent shells are that we are not supporting...

Comment (Naatan) [reply internal/subshell/subshell.go:123]: Good idea! Please file a ticket :)

**Comment by Naatan** [comment]
> Maybe we should support it

Sounds like a good idea, especially if it's the default on Ubuntu. Please file a ticket :)

File: internal/subshell/subshell.go [hunk 3]
```diff
@@ -84,10 +86,20 @@ func New() SubShell {
...
-	logging.Debug("Detected SHELL: %s", binary)
+	// try to find the binary on the PATH
+	binaryPath, err := exec.LookPath(binary)
+	if err == nil {
+		// if we found it, resolve all symlinks, for many Linux distributions the SHELL is "sh" but symlinked to a different default shell like bash or zsh
+		resolved, err := fileutils.ResolvePath(binaryPath)
+		if err != nil {
+			logging.Debug("Failed to resolve path to shell binary %s: %v", binaryPath, err)
+		}
+		binary = resolved
```
Comment (Naatan) [inline internal/subshell/subshell.go:97]: This should probably be in an `else`, since the value of `resolved` is not guaranteed if `err == nil`.

**Review by Naatan — CHANGES_REQUESTED** [review]

File: internal/subshell/subshell.go (hunk 3 above)
Comment (mdrohmann) [reply internal/subshell/subshell.go:97]: Good point.

File: internal/subshell/subshell.go (hunk 2 above)
Comment (mdrohmann) [reply internal/subshell/subshell.go:124]: https://www.pivotaltracker.com/story/show/177629477

**Review by Naatan — APPROVED** [review]

**Pull request merged**